File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        res = 0
        for i in range(len(heights)):
            start = i
            while(stack and stack[-1][1] > heights[i]):
                s,h = stack.pop()
                res = max(res, h*(i-s))
                start = s
            stack.append((start,heights[i]))
        while(stack):
            s,h = stack.pop()
            res = max(res, h*(len(heights)-s))
        return res

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<int[]> stack = new Stack();
        int res = 0;
        for(int i = 0; i<heights.length; i++){
            int s = i;
            while(!stack.isEmpty() && stack.peek()[1] > heights[i]){
               int [] top = stack.pop();
               res = Math.max( res,top[1] * (i - top[0]) );
               s = top[0];
            }
            stack.push(new int[] {s,heights[i]});
        }
        while(!stack.isEmpty()){
            int [] top = stack.pop();
            res = Math.max(res, top[1]*(heights.length - top[0]) );
        }
        return res;
    }
}